invisible_cities/cities/detsim_waveforms.py

```diff
@@ -0,0 +1,76 @@
+import numpy as np
+
+##################################
+######### WAVEFORMS ##############
+##################################
+def create_waveform(times : np.ndarray,
+                    pes   : np.ndarray,
+                    bins  : np.ndarray,
+                    nsamples : int) -> np.ndarray:
+    """
+    This function builds a waveform from a set of (buffer_time, pes) values.
+    This set is of values come from the times and pes arguments.
+
+    Input
+    times: a vector with the buffer times in which a photoelectron is produced in the
+    detector.
+    pes (from photoelecton): a vector with the photoelectrons produced in the detector in
+    each of the buffer times in times argument.
+    bins: a vector with the output waveform bin times (for example [0, 25, 50, ...] if
+    the detector has a sampling time of 25).
+    nsamples: an integer that controlls the distribution of the photoelectrons in each of
+    the waveform bins. The counts (N) in a given time bin (T) are distributed in the waveform
+    such as the nsamples posterior to T would have N/nsamples counts (included T).
+    """
+    
+    wf = np.zeros(len(bins))
+
+    if np.sum(pes)==0:
+        return wf
+    if nsamples==0:
+        t = np.repeat(times, pes)
+        sel = (bins[0]<=t) & (t<=bins[-1])
+        t = t[sel]
+        indexes = np.digitize(t, bins)-1
+        indexes, counts = np.unique(indexes, return_counts=True)
+        wf[indexes] = counts
+        return wf
+
+    t = np.repeat(times, pes)
+    sel = (bins[0]<=t) & (t<=bins[-1])
+    t = np.clip  (t[sel], bins[0], bins[-nsamples])
+
+    indexes = np.digitize(t, bins)-1
+    indexes, counts = np.unique(indexes, return_counts=True)
+
+    spread_counts = np.repeat(counts[:, np.newaxis]/nsamples, nsamples, axis=1)
+    for index, counts in zip(indexes, spread_pes):
+        wf[index:index+nsamples] += counts
+    return wf
+
+
+def create_sensor_waveforms(times   : np.ndarray,
+                            pes_at_sensors : np.ndarray,
+                            wf_buffer_time : float,
+                            bin_width    : float,
+                            nsamples : int,
+                            poisson  : bool =False) -> np.ndarray:
+    """
+    This function calls recursively to create_waveform. See create_waveform for
+    an explanation of the arguments not explained below.
+
+    Input:
+    pes_at_sensors: an array with size (#sensors, len(times)). It is the same
+    as pes argument in create_waveform but for each sensor in axis 0.
+    wf_buffer_time: a float with the waveform extent (in default IC units)
+    bin_width: a float with the time distance between bins in the waveform buffer.
+    poisson: a bool. This must be set to True to poisson distribute each bin count.
+    """
+
+    bins = np.arange(0, wf_buffer_time, bin_width)
```

Comment:
actually, because these are bin edges, you need to have one more entry. Otherwise your waveform goes from 0 to `wf_buffer_time - bin_width`
```suggestion
    bins = np.arange(0, wf_buffer_time + bin_width, bin_width)
```
which means that in `create_waveforms` you need to do `wf = np.zeros(len(bins) - 1)`

Reply:
`bins` are bin centers instead of bin edges

Reply:
No, they are bin edges.
If they were bin centers, you would have a bin center that is zero (unphysical) and
```
sel = (bins[0]<=t) & (t<=bins[-1])
```
and
```
np.digitize(t, bins)
```
would be wrong.
Besides, everything is easier if you work with bin edges whenever possible. It takes a bit of a mental switch for us but it is worth it.

Reply:
I think it it better to think in bin edges, so I changed it as you say. Nevertheless, `np.digitize` can be think to be working between bin centers as I was thinking in it before, which I know it is less natural.

I changed the distribution method. Now, the waveform is expanded to distribute the counts out of the real waveform. This is more natural since in this way we avoid clipping, which accumulate counts in a single bin. In this way, count distribution in nsamples bins is now continuos, in the sense that if we attach another waveform to this one, the distribution of bins would be maintaining its nsamples distribution correct. At the end, we shrink the waveform to the real size.

Note that commit 5b41aef is wrong, and it is corrected before.